File: src/2020/day8/day8.cc
#include <fstream>
#include <iostream>
#include <sstream>
#include <map>
#include <string>
#include <vector>

using namespace std;
struct instruction {
  string opcode;
  int operand;
  bool visited;
};

int main() {
  fstream file("2020_day8.txt");
  vector<instruction> instructions;



  string line;
  stringstream s;

  string opcode;
  int operand;
  bool visited;
  visited = false;

  while (getline(file, line)) {
    s = stringstream(line);
    s >> opcode >> operand;
    instructions.emplace_back(instruction{opcode, operand, visited});
  }
    
  int accumulator = 0;
  int ip = 0;
  while(true){
   instruction& instruction_i=  instructions[ip];
    if (instruction_i.visited){
      break;
    }else{
      instruction_i.visited = true;
    }
    if (instruction_i.opcode == "nop"){
    }else if (instruction_i.opcode == "acc"){
      accumulator += instruction_i.operand;
    }
    else if (instruction_i.opcode == "jmp"){
      ip +=  instruction_i.operand;
      continue;
    }
    ip++;

  }
  cout << accumulator << endl;
  return 0;
}

File: src/2020/day8/day8.py
import time
with open("2020_day8.txt", "r") as f:
	lines = f.read().split("\n")
def get_acc(lines, visited):
	accumulator = 0
	ip = 0
	while True:
		if visited[ip] == True:break
		visited[ip] = True
		line = lines[ip]
		instruction = line[0]
		arguemnt =int(line[1])
		if (instruction == "nop"):
			pass
		elif(instruction == "acc"):
			accumulator += arguemnt
		elif(instruction == "jmp"):
			ip += arguemnt
			continue

		ip+= 1

	return accumulator

visited = [False for line in lines]
lines = [line.split() for line in lines]

st = time.time()
accumulator = get_acc(lines, visited)
print(time.time() - st)
print(accumulator)

